source/vibe/web/rest.d

```diff
@@ -762,336 +613,253 @@ private HTTPServerRequestDelegate jsonMethodHandler(alias Func, size_t ridx, T)(
-				static assert (
-					PTT.length == 0,
-					"Interface getters may not have parameters."
-				);
-
-				if (!tps.canFind(RT.stringof)) {
-					tps ~= RT.stringof;
-					string implname = RT.stringof ~ "Impl";
-
-					ret ~= q{
-						m_%s.requestFilter = m_requestFilter;
-					}.format(implname);
-					ret ~= "\n";
+	InetHeaderMap headers;
+	InetHeaderMap reqhdrs;
+	InetHeaderMap opthdrs;
+
+	string url_prefix;
+
+	auto query = appender!string();
+	auto jsonBody = Json.emptyObject;
+	string body_;
+
+	void addQueryParam(size_t i)(string name)
+	{
+		if (query.data.length) query.put('&');
+		query.filterURLEncode(name);
+		query.put("=");
+		static if (is(PT == Json))
+			query.filterURLEncode(ARGS[i].toString());
+		else // Note: CTFE triggers compiler bug here (think we are returning Json, not string).
+			query.filterURLEncode(toRestString(serializeToJson(ARGS[i])));
+	}
+
+	foreach (i, PT; PTT) {
+		enum sparam = sroute.parameters[i];
+		auto fieldname = route.parameters[i].fieldName;
+		static if (sparam.kind == ParameterKind.query) {
+			addQueryParam!i(fieldname);
+		} else static if (sparam.kind == ParameterKind.body_) {
+			jsonBody[fieldname] = serializeToJson(ARGS[i]);
+		} else static if (sparam.kind == ParameterKind.header) {
+			// Don't send 'out' parameter, as they should be default init anyway and it might confuse some server
+			static if (sparam.isIn) {
+				static if (isInstanceOf!(Nullable, PT)) {
+					if (!ARGS[i].isNull)
+						headers[fieldname] = to!string(ARGS[i]);
+				} else headers[fieldname] = to!string(ARGS[i]);
+			}
+			static if (sparam.isOut) {
+				// Optional parameter
+				static if (isInstanceOf!(Nullable, PT)) {
+					opthdrs[fieldname] = null;
+				} else {
+					reqhdrs[fieldname] = null;
 				}
 			}
 		}
 	}
-	return ret;
-}
 
-private mixin template RestClientMethods(I) if (is(I == interface)) {
-	mixin RestClientMethods_MemberImpl!(__traits(allMembers, I));
-}
+	debug body_ = jsonBody.toPrettyString();
+	else body_ = jsonBody.toString();
+
+	string url;
+	foreach (i, p; route.pathParts) {
+		if (p.isParameter) {
+			switch (p.text) {
+				foreach (j, PT; PTT) {
+					case sroute.parameters[j].name:
+						url ~= urlEncode(toRestString(serializeToJson(ARGS[j])));
+						goto sbrk;
+				}
+				default: url ~= ":" ~ p.text; break;
+			}
+			sbrk:;
+		} else url ~= p.text;
+	}
 
-// Poor men's `foreach (method; __traits(allMembers, I))`
-// The only way to emulate a foreach in a mixin template is to mixin a recursion
-// of that template.
-private mixin template RestClientMethods_MemberImpl(Members...) {
-	import std.traits : MemberFunctionsTuple;
-	static assert (Members.length > 0);
-	// WORKAROUND #1045 / @@BUG14375@@
-	static if (Members[0].length != 0) {
-		private alias Ovrlds = MemberFunctionsTuple!(I, Members[0]);
-		// Members can be declaration / fields.
-		static if (Ovrlds.length > 0) {
-			mixin RestClientMethods_OverloadImpl!(Ovrlds);
+	scope (exit) {
```

Comment:
Since this code existed before this PR, I'm going to open a separate ticket for this.